### PR TITLE
Renaming file should rename class

### LINKED ISSUE #109: Project System: Renaming file should rename class
_From @davkean on March 24, 2016 17:16_

In csproj/msvbproj, renaming a file updates the first class within that file if root namespace + folder matches the full namespace.

We should enable that for parity, but we should also be a little smarter around this:

- [x] Currently, csproj/msvbproj only does this if the full namespace matches the folder, we shouldn't require this.

- [x] Prompt user for rename based on option

_Copied from original issue: dotnet/roslyn#10064_

### PR DESCRIPTION
**Pull request by jinujoseph**
fixes  #109.

Currently since there isn't an event to listen which indicates a rename of file completion,  Listening to workspacechange event. Will have to come back and put the right events.

This addresses the first requirement that the namespace needn't have to follow folder structure.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
+﻿//--------------------------------------------------------------------------------------------
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: Wrong copyright header. Also other new files are missing copyright header.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+// FileRenameTracker
+//
+// Exports an IProjectChangeHintReceiver to listen to file renames. If the file being renamed
+// is a code file, it will prompt the user to rename the class to match. The rename is done
+// using code model
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: Not done using codemodel anymore right?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                {
+                    var kvp = hint.RenamedFiles.First();
+                    await ScheduleRenameAsync(kvp.Key, kvp.Value).ConfigureAwait(false);
+                }
+            }
+        }
+
+        public Task HintingAsync(IProjectChangeHint hint)
+        {
+            return TplExtensions.CompletedTask;
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: You can just use `Task.CompletedTask`.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        private readonly IUnconfiguredProjectVsServices _unconfiguredProjectVsServices;
+        private IComponentModel _componentModel;
+        private VisualStudioWorkspace _visualStudioWorkspace;
+
+        /// <summary>
+        /// The thread handling service.
+        /// </summary>
+        [Import]
+        private IProjectThreadingService ThreadingService { get; set; }
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: You can get this from `_unconfiguredProjectVsServices.ThreadingService` instead of importing it.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            {
+                return;
+            }
+
+            await ThreadingService.SwitchToUIThread();
+
+            if (_visualStudioWorkspace == null)
+            {
+                _componentModel = (IComponentModel)ServiceProvider.GetService(typeof(SComponentModel));
+                _visualStudioWorkspace = _componentModel.GetService<VisualStudioWorkspace>();
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: Curious. Why not just import it? Is there some lifetime issue? If so please add a comment.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+                _visualStudioWorkspace.WorkspaceChanged -= OnWorkspaceChanged;
+
+                _threadingService.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    var myNewProject = _visualStudioWorkspace.CurrentSolution.Projects.Where(p => string.Equals(p.FilePath, _project.FilePath, StringComparison.OrdinalIgnoreCase)).FirstOrDefault();
+                    Document newDocument = (from d in myNewProject.Documents where d.FilePath.Equals(_newFilePath, StringComparison.OrdinalIgnoreCase) select d).FirstOrDefault();
+                    if (newDocument == null)
+                        return;
+
+                    var root = newDocument.GetSyntaxRootAsync().Result;
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: await?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                        return;
+
+                    string oldName = Path.GetFileNameWithoutExtension(_oldDocument.FilePath);
+                    string newName = Path.GetFileNameWithoutExtension(newDocument.FilePath);
+
+                    var declaration = root.DescendantNodes().Where(n => HasMatchingSyntaxNode(n, oldName)).FirstOrDefault();
+                    if (declaration == null)
+                        return;
+
+                    var semanticModel = newDocument.GetSemanticModelAsync().Result;
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: await

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                            if (result == (int)VSConstants.MessageBoxResult.IDNO)
+                            {
+                                return;
+                            }
+                        }
+                    }
+
+                    // Now do the rename
+                    var optionSet = newDocument.Project.Solution.Workspace.Options;
+                    var renamedSolution = RoslyRenamer.Renamer.RenameSymbolAsync(newDocument.Project.Solution, symbol, newName, optionSet).Result;
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: await

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+                                                OLEMSGBUTTON.OLEMSGBUTTON_OK, OLEMSGDEFBUTTON.OLEMSGDEFBUTTON_FIRST);
+                    }
+                });
+
+            }
+        }
+
+        private bool HasMatchingSyntaxNode(SyntaxNode syntaxNode, string name)
+        {
+            if (syntaxNode is BaseTypeDeclarationSyntax || syntaxNode is DelegateDeclarationSyntax)
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: This needs to be replicated for VB. Instead I suggest using SyntaxGenerator here so that it can remain language agnostic. Something like:

```csharp
var generator = SyntaxGenerator.GetGenerator(document);
var kind = generator.GetDeclarationKind(syntaxNode);

if (kind == DeclarationKind.Class || DeclarationKind.Interface || ...) 
{
     return generator.GetName(syntaxNode) == name;
}
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs (hunk 4 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: Ah , thanks ...fixed it

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs (hunk 8 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: oh yes , all of this should be called async ..fixed it

**Comment by davkean** [comment]
You want to sync with the latest changes.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    [Export(typeof(IProjectChangeHintReceiver)), Export]
+    [ProjectChangeHintKind(ProjectChangeFileSystemEntityRenameHint.RenamedFileAsString)]
+    [AppliesTo(ProjectCapability.CSharp)]
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: Why only C#?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private IComponentModel _componentModel;
+        private VisualStudioWorkspace _visualStudioWorkspace;
+
+        /// <summary>
+        /// Gets the VS global service provider.
+        /// </summary>
+        [Import]
+        protected SVsServiceProvider ServiceProvider { get; private set; }
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: Use constructor injection (ie import via constructor instead of via fields)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs (hunk 10 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: Have not tested in VB yet :-) ...will enable once I run few tests ..

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IServiceProviderExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+            where ServiceType : class
+        {
+            InterfaceType service = null;
+
+            try
+            {
+                service = sp.GetService(typeof(ServiceType)) as InterfaceType;
+            }
+            catch
+            {
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IServiceProviderExtensions.cs]: Change you change this to 1) not use an "as" cast and 2) not to catch exceptions. Both of these hide bugs.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+                    if (semanticModel == null)
+                        return;
+
+                    var symbol = semanticModel.GetDeclaredSymbol(declaration);
+                    if (symbol == null)
+                        return;
+
+                    // We're about to do a symbolic rename.If the user has asked us to prompt, We need to open a dialog and ask them  
+                    //  Otherwise go ahead and do the rename.
+                    EnvDTE.DTE dte = _serviceProvider.GetService<EnvDTE.DTE, EnvDTE.DTE>();
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: Does below this have to be done on the UI thread?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+                    var semanticModel = await newDocument.GetSemanticModelAsync().ConfigureAwait(false);
+                    if (semanticModel == null)
+                        return;
+
+                    var symbol = semanticModel.GetDeclaredSymbol(declaration);
+                    if (symbol == null)
+                        return;
+
+                    // We're about to do a symbolic rename.If the user has asked us to prompt, We need to open a dialog and ask them
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: Below this line should be done on the UI thread

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Shell/Interop/VsHierarchyExtensions.cs [hunk 15]
```diff
@@ -54,5 +54,18 @@ public static T GetProperty<T>(this IVsHierarchy hierarchy, HierarchyId item, Vs
             // NOTE: We consider it a bug in the underlying project system or the caller if this cast fails
             return (T)resultObject;
         }
+                
+        /// <summary>
+        /// Returns EnvDTE.Project object for the hierarchy
+        /// </summary>
+        public static EnvDTE.Project GetDTEProject(this IVsHierarchy hierarchy)
+        {
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Shell/Interop/VsHierarchyExtensions.cs]: Can you replace usage of this method with the generic one above it? I'm not a fan of the overly defensive casts, and HRESULT checks - as they hide bugs. It would a bug if our hierarhcy didn't have a extensibility object that didn't implement EnvDTE.Project

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+using System;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using RoslyRenamer = Microsoft.CodeAnalysis.Rename;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: "Roslyn"

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+            }
+
+            if (args.Kind == WorkspaceChangeKind.DocumentChanged && args.ProjectId == _project.Id)
+            {
+                _docChanged = true;
+            }
+
+            if (_docAdded && _docRemoved && _docChanged)
+            {
+                _visualStudioWorkspace.WorkspaceChanged -= OnWorkspaceChanged;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: So we only unsubscribe if these checks are true? Does this mean that we never unsubscribe in some cases?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+                _threadingService.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    var myNewProject = _visualStudioWorkspace.CurrentSolution.Projects.Where(p => string.Equals(p.FilePath, _project.FilePath, StringComparison.OrdinalIgnoreCase)).FirstOrDefault();
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: You're reading some things here, such as project.FilePath on a background thread.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs (hunk 5 above)
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: Ah, VisualStudioWorkspace is a MEF component, yes we should import it instead of having this IServiceProvider dance.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs [hunk 19]
```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+//--------------------------------------------------------------------------------------------
+// FileRenameTracker
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: We put comments like this as XML comments.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+{
+    internal class Renamer
+    {
+        private readonly VisualStudioWorkspace _visualStudioWorkspace;
+        private readonly IProjectThreadingService _threadingService;
+        private readonly SVsServiceProvider _serviceProvider;
+        private readonly Project _project;
+        private readonly Document _oldDocument;
+        private readonly string _newFilePath;
+        private bool _docAdded = false;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: Why are these local fields instead of just local variables?

**Comment by davkean** [comment]
This is a great start, a couple of additional comments:

1) You want to make sure that any non-CPS/Roslyn API that you call is done on a UI thread

2) I think you'll want a couple of interfaces/components in-between that make testing this thing easy. At the moment, you're going to have to figure out how to mock VisualStudioWorkspace to unit test this.

3) Be mindful of any mutable state that you have - every component in CPS is free-threaded, any state that you mutate you need make sure is thread-safe.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+
+            if (_visualStudioWorkspace == null)
+            {
+                _componentModel = (IComponentModel)ServiceProvider.GetService(typeof(SComponentModel));
+                _visualStudioWorkspace = _componentModel.GetService<VisualStudioWorkspace>();
+            }
+
+            IVsHierarchy hierarchy = _projectVsServices.Hierarchy;
+
+            EnvDTE.Project project = hierarchy.GetDTEProject();
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: You can get the project.FullName from _projectVsServices.Project.FullPath

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs (hunk 18 above)
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: You can get this from UnconfiguredProject.FullPath (_services.Project.FullPath)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+using System.Linq;
+using Task = System.Threading.Tasks.Task;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    [Export(typeof(IProjectChangeHintReceiver)), Export]
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: You don't need this second [Export]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+            _oldDocument = (from d in project.Documents where d.FilePath.Equals(oldFilePath, StringComparison.OrdinalIgnoreCase) select d).FirstOrDefault();
+        }
+
+        public void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs args)
+        {
+            if (args.Kind == WorkspaceChangeKind.DocumentAdded && args.ProjectId == _project.Id)
+            {
+                Project project = (from p in args.NewSolution.Projects where p.Id.Equals(_project.Id) select p).FirstOrDefault();
+                Document addedDocument = (from d in project.Documents where d.Id.Equals(args.DocumentId) select d).FirstOrDefault();
+                if (addedDocument.FilePath.Equals(_newFilePath, StringComparison.OrdinalIgnoreCase))
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: StringComparers.Path.Equals( )

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs (hunk 20 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: As discussed leaving this as is.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs (hunk 17 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: Thatz correct ,

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs (hunk 5 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: Ah nice , did not knew :-)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Shell/Interop/VsHierarchyExtensions.cs (hunk 15 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Shell/Interop/VsHierarchyExtensions.cs]: Sure , done

**Comment by jinujoseph** [comment]
I added few Unit Tests and did bit of refactoring to make the testable.
pls review when you have a moment

**Comment by davkean** [comment]
Sorry, I ended up in meetings all afternoon. Will look at this tomorrow morning.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs [hunk 24]
```diff
@@ -1,44 +1,42 @@
...
-        /// </summary>
-        [Import]
-        protected SVsServiceProvider ServiceProvider { get; private set; }
-        
         [ImportingConstructor]
-        public FileRenameTracker(IUnconfiguredProjectVsServices projectVsServices)
+        public FileRenameTracker(IUnconfiguredProjectVsServices projectVsServices, VisualStudioWorkspace visualStudioWorkspace, SVsServiceProvider serviceProvider)
         {
             Requires.NotNull(projectVsServices, nameof(projectVsServices));
+            Requires.NotNull(visualStudioWorkspace, nameof(VisualStudioWorkspace));
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs]: Nit: These are supposed to be names of the "parameters" not the types

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs [hunk 25]
```diff
@@ -107,14 +110,12 @@ public void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs args)
...
-                    }
+                    }                    
 
                     // Now do the rename
                     var optionSet = newDocument.Project.Solution.Workspace.Options;
-                    var renamedSolution = await RoslyRenamer.Renamer.RenameSymbolAsync(newDocument.Project.Solution, symbol, newName, optionSet).ConfigureAwait(false);
-
-                    await _threadingService.SwitchToUIThread();
-
+                    var renamedSolution = await RoslynRenamer.Renamer.RenameSymbolAsync(newDocument.Project.Solution, symbol, newName, optionSet).ConfigureAwait(true);
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs]: Can you add a comment here that TryApplyChanges needs to occur on the UI thread? It's not obvious, for example, I could move this RoslynRenamer call to a different location, not knowing that I've just changed behavior to the next call.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IVsEnvironmentServices.cs [hunk 26]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    internal interface IVsEnvironmentServices
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IVsEnvironmentServices.cs]: Can we change this a little to:

1) Not include start with IVs - we reserve these for types that actually come from VS.
2) Be a little more isolated, or more specific to rename. This looks like a grab bag of VS functionality on things you just wanted to isolate. For example, display errors to the user to completely different concept to applying changes to Roslyn. They shouldn't be tied to each other.

**Comment by davkean** [comment]
I went through the new commits one-by-one and added comments, sorry if I added comments to no longer existent/changed code.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IVsEnvironmentServices.cs (hunk 26 above)
Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IVsEnvironmentServices.cs]: ya , i had a feeling about that as well . does IEnvironementService sounds good i was not very happy with that as well ? and i can create IRenamer to do things specific to rename.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/OptionsSettings.cs [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Composition;
+using Microsoft.VisualStudio.Shell;
+
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    /// <summary>
+    ///     Provides an implementation of <see cref="IOptionsSettings"/> that delegates onto
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/OptionsSettings.cs]: Comment not finished.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IUserNotificationServices.cs [hunk 28]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    internal interface IUserNotificationServices
+    {
+        Task<bool> CheckPromptAsync(string promptMessage);
+
+        void NotifyFailureAsync(string renamedString);
+
+        Task<bool> CheckPromptForRenameAsync(string promptMessage);
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IUserNotificationServices.cs]: The methods that are associated with "rename" don't really belong here - as this is a general service unrelated to rename. However, given below - they are very simple methods that delegate on to other abstractions, is there any reason you need to mock them? ie, Why isn't CheckPromptForRenameAsync just a mock of CheckPromptAsync and IOptionsSettings?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IUserNotificationServices.cs [hunk 29]
```diff
@@ -0,0 +1,17 @@
...
+
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    internal interface IUserNotificationServices
+    {
+        Task<bool> CheckPromptAsync(string promptMessage);
+
+        void NotifyFailureAsync(string renamedString);
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IUserNotificationServices.cs]: This isn't async, and the parameter name should be changed to better reflect it's purpose.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RoslynServices.cs [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+        public async Task<Solution> RenameSymbolAsync(Solution solution, ISymbol symbol, string newName)
+        {
+            var optionSet = solution.Workspace.Options;
+            return await RoslynRenamer.Renamer.RenameSymbolAsync(solution, symbol, newName, optionSet).ConfigureAwait(false);
+        }
+        /// <summary>
+        ///  Applies the provided solution to the workspace 
+        ///  Make sure this is always called from an UI Thread.
+        /// </summary>
+        /// <param name="ws"></param>
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RoslynServices.cs]: Remove the parameters if you're not going to fill them in.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UserNotificationServices.cs [hunk 31]
```diff
@@ -2,30 +2,31 @@
...
 using Microsoft.VisualStudio.Composition;
 using Microsoft.VisualStudio.Shell.Interop;
 using Microsoft.VisualStudio.Shell;
-using RoslynRenamer = Microsoft.CodeAnalysis.Rename;
 
 namespace Microsoft.VisualStudio.ProjectSystem.VS
 {
     /// <summary>
-    ///     Provides an implementation of <see cref="IVsEnvironmentServices"/> that delegates onto 
+    ///     Provides an implementation of <see cref="IUserNotificationServices"/> that delegates onto
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UserNotificationServices.cs]: Comment not finished.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UserNotificationServices.cs [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+            if (result == (int)VSConstants.MessageBoxResult.IDNO)
+            {
+                return false;
+            }
+            return true;
+        }
+
+        public async void NotifyFailureAsync(string failureMessage)
+        {
+            await _threadingService.SwitchToUIThread();
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UserNotificationServices.cs]: We shouldn't have async voids as they aren't observable, and exceptions go unhandled.

I'm also unconvinced by this pattern of having "async" methods that are really just pumping to the UI thread. I think we should change the notification service to be synchronous and have the caller worry about switching to the UI thread. We should have such services check to make sure they are called on the UI thread, I'll do some research as CPS might already have some helper classes for types that are not free-threaded.

**Comment by davkean** [comment]
:+1: 

Can you make sure there's items/bugs on:

1) Adding VB support
2) Extending the pattern (generics, etc)
3) Moving from workspace changes -> language service notification
4) Introducing a base class that verifies something is in UI thread.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/OptionsSettings.cs [hunk 33]
```diff
@@ -8,27 +8,21 @@
...
-            Requires.NotNull(threadingService, nameof(threadingService));
             _serviceProvider = serviceProvider;
-            _threadingService = threadingService;
         }
 
-        public async Task<T> GetPropertiesValueAsync<T>(string category, string page, string property, T defaultValue)
+        public T GetPropertiesValue<T>(string category, string page, string property, T defaultValue)
         {
-            await _threadingService.SwitchToUIThread();
+            //TODO : VerifyOnUIThread();
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/OptionsSettings.cs]: You can actually verify on UI thread via the `IProjectThreadingService.VerifyOnUIThread()` method.

Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/OptionsSettings.cs]: ya , thatz what i thought but the method has not been implemented yet correct ?

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/OptionsSettings.cs]: I was looking for a base class that already provided this feature - but given the base class will just import IProjectThreadingService and call this method it's not going to add any value.

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/OptionsSettings.cs]: ie, This method is the one to use and it works.

Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/OptionsSettings.cs]: cool , i will add that as separate PR and create bugs/items for the rest

Comment (jinujoseph) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/OptionsSettings.cs]: filed #229 #227 #228
